syntax = "proto3";

package helper;

import "protoc-gen-openapiv2/options/annotations.proto";

service STHelper {
  rpc AnalyzeFinancialData (FinancialRequest) returns (FinancialResponse) {}
  rpc GetStockQuote (StockQuoteRequest) returns (StockQuoteResponse) {}
  rpc GetHistoricalStockData (HistoricalStockDataRequest) returns (HistoricalStockDataResponse) {}
}

message FinancialRequest {
  string userId = 1;
  string financial_data = 2;
}

message FinancialResponse {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      type: STRING,
    }
  };
  string suggestions = 1;
}

message StockQuoteRequest {
  string symbol = 1;
  string provider = 2;
}

message StockQuoteResponse {
  string symbol = 1;
  string name = 2;
  string exchange = 3;
  float lastPrice = 4;
  float open = 5;
  float high = 6;
  float low = 7;
  int64 volume = 8;
  float prevClose = 9;
  float change = 10;
  float changePercent = 11;
  float yearHigh = 12;
  float yearLow = 13;
  int64 marketCap = 14;
  int64 sharesOutstanding = 15;
  float pe = 16;
  string earningsAnnouncement = 17;
  float eps = 18;
  string sector = 19;
  string industry = 20;
  float beta = 21;
}

message HistoricalStockDataRequest {
  string symbol = 1;
  string start_date = 2;
  string end_date = 3;
  string provider = 4;
}

message HistoricalStockDataResponse {
  repeated float open = 1;
  repeated float high = 2;
  repeated float low = 3;
  repeated float close = 4;
  repeated int64 volume = 5;
  repeated float vwap = 6;
  repeated float adj_close = 7;
  repeated float unadjusted_volume = 8;
  repeated float change = 9;
  repeated float change_percent = 10;
}